### PR TITLE
[py12] ml cpu tests

### PR DESCRIPTION
**Pull request by can-anyscale**
Add ml-cpu tests for python 3.12. Skip those tests that are relying on tensorflow, we won't suppport tensorflow for python 3.12 (for now).

Test:
- CI

### CODE REVIEW

File: ci/docker/serve.build.Dockerfile [hunk 1]
```diff
@@ -26,12 +26,10 @@ pip install -U --ignore-installed \
   -r python/requirements.txt \
   -r python/requirements/test-requirements.txt
 
-# TODO(can): upgrade tensorflow for python 3.12
-if [[ "${PYTHON-}" != "3.12" ]]; then
-  pip install -U -c python/requirements_compiled.txt \
-    tensorflow tensorflow-probability torch torchvision \
-    transformers aioboto3
-fi
+pip install -U -c python/requirements_compiled.txt \
```
Comment (can-anyscale) [inline ci/docker/serve.build.Dockerfile:29]: do not need this now that we install tensorflow (but not really using it) in the image

File: python/ray/tune/tests/test_sample.py [hunk 2]
```diff
@@ -1846,6 +1879,10 @@ def testConstantGridSearchBasicVariant(self):
         self.assertNotEqual(configs[0]["rand"], configs[3]["rand"])
 
     @patch.object(logger, "warning")
+    @pytest.mark.skipif(
+        sys.version_info >= (3, 12),
+        reason="TODO(justinvyu): not working for python 3.12 yet",
```
Comment (can-anyscale) [inline python/ray/tune/tests/test_sample.py:1884]: @justinvyu: all tests are passing now after I skip the tensorflow related test, except for this one https://buildkite.com/ray-project/postmerge/builds/5637#0190ef5b-9819-4e7a-9f53-9fdb6639b5de/176-2505. I'm jus skipping this for now but if you can help take a look that would be awesome, thankks

File: .buildkite/ml.rayci.yml [hunk 3]
```diff
@@ -40,6 +52,29 @@ steps:
...
+        --python-version {{matrix.python}}
+        --except-tags gpu_only,gpu,minimal,doctest,needs_credentials,soft_imports,rllib,multinode
+    depends_on: 
+      - mlbuild-multipy
+      - forge
+    job_env: 
+    matrix:
+      setup:
+        python: ["3.12"]
+        worker_id: ["0", "1", "2", "3"]
```
Comment (aslonnie) [inline .buildkite/ml.rayci.yml:76]: why we need this kind of matrix here?

**Review by aslonnie — COMMENTED** [review]
this PR looks really big. is there possibility to split this into smaller ones that are easier to reason about?

**Comment by can-anyscale** [comment]
@aslonnie the bulk of this is to skip tests for python 3.12 and mainly for the ml folks to review ;)

File: .buildkite/ml.rayci.yml (hunk 3 above)
Comment (can-anyscale) [reply .buildkite/ml.rayci.yml:76]: it's the number of build-kite jobs required, computed by the number of python versions x number of test shards required for each python

```
shard/python   3.10  3.11  3.12
0               X.      X.     X
1                          ...
2
3
```

**Pull request closed without merging**